WIP: how-to guide for contributing a new expectation and expectation templates (#2283)
great-expectations/great_expectations: docs/guides/how_to_guides/creating_and_editing_expectations/how_to_contribute_a_new_expectation_to_great_expectations.rst

.. _how_to_guides__creating_and_editing_expectations__how_to_template:

How to contribute a new Expectation to Great Expectations
==========================================================

This guide will help you add a new Expectation to Great Expectations’ shared library. Your Expectation will be featured in the Expectations Gallery, along with many others developed by data practitioners from around the world as part of this collaborative community effort.

.. admonition:: Prerequisites: This how-to guide assumes you have already:

  - :ref:`Set up your dev environment to contribute <contributing_setting_up_your_dev_environment>`
  - :ref:`Signed the Contributor License Agreement (CLA) <contributing_cla>`

Steps
-----

#. Choose a Parent Class to Help your Implementation.

>     There are four Expectation classes that make the development of particular types of Expectations significantly easier by hiding the "crud" code and letting you focus on the business logic of your Expectation. Decide which one suites your Expectation:

Reviewer: ```suggestion
    There are four Expectation subclasses that make the development of particular types of Expectations significantly easier by handling boilerplate code and letting you focus on the business logic of your Expectation. Consider choosing one that suites your Expectation:
```

                The convention is to implement a new Metric Provider (a class that can compute a metric) that your Expectation depends on in the same file as the Expectation itself.

                Search for ``class ColumnValuesEqualThree`` and rename it to ColumnValues<CamelCase version of the second part of the metric name that you declared in the previous step>.

                The Metric Provider class declares the condition metric that it can compute. "Condition metric" is a metric that answers a yes/no question:

                .. code-block:: python

                    condition_metric_name = "column_values.equal_three"

                The parent class expects the variable ``condition_metric_name`` to be set. Change the value of ``condition_metric_name`` to the same name that you used for ``map_metric`` in your Expectation class.

                The Expectation declares that it needs a yes/no Metric "X" and the Metric Provider declares that it can compute this Metric. A match made in heaven.

>             * Implement the computation of the Metric in your new Metric Provider class for at least one of the three backends (Execution Engines) that Great Expectations supports: pandas, sqlalchemy, spark. Most contributors find starting with Pandas is the easiest and fastest way to build.

Reviewer: ```suggestion
            * Implement the computation of the Metric in your new Metric Provider class for at least one Execution Engines that Great Expectations supports, such as pandas, sqlalchemy, or spark. Most contributors find that starting with Pandas is the easiest and fastest way to build.
```

                The ``engine`` argument of ``column_condition_partial`` is set to ``PandasExecutionEngine`` to signal to the method in the parent that the method computes the Metric for pandas backend.
                There is nothing special about the name of the method ``_pandas`` - it can be called anything else, but why make things more complicated than they must be?

                Implement this method to compute your Metric.

                .. admonition:: Note:

                    How to support additional arguments your Expectation needs.

                    The Expectation in the template (``expect_column_values_to_equal_three``) did not need to accept any additional arguments to evaluate the data.

                    Here is how you could modify ``expect_column_values_to_equal_three`` to ``expect_column_values_to_equal_integer``, where users would have to specify the value of the integer as an argument:

                    * Find the snippet ``success_keys = ("mostly",)`` in the class that implements your Expectation. Add your arguments to ``success_keys``

                    .. code-block:: python

>                         success_keys = ("mostly", "integer")

Reviewer: These are ordered (sort of ... in practice users just need to be explicit in naming kwargs), but I think integer should come first to maintain the usual signature where mostly is at the end.

Author: +1

                    Success keys are arguments that determine the values of the Expectation's metrics and when the Expectation will succeed.

                    * In the class that implements Metric Provider set the variable ``condition_value_keys`` to a tuple of your arguments:

> 

Reviewer: Something like this to be added:

"value_keys" work for Metrics like "success_keys" do for Expectations, but they are used to determine the *value* of the metric (hence the name!). If your metric needs additional user-supplied parameters, you add them to the value_keys.

For a map Metric producing a yes/no question answer, you use "condition_value_keys" (because it's the *condition* part of the metric).

Author: Adding...

                    This requires defining a new Metric. The parent class of your Metric Provider class (``ColumnMapMetricProvider``) provides support for this case.

                    A good example of this pattern is ``expect_column_value_z_scores_to_be_less_than`` - one of the core Expectations.

                    The Expectation declares  "column_values.z_score.under_threshold" as its ``condition_metric_name`` (the Metric that answers the yes/no question for every row).

                    The ``ColumnValuesZScore`` Metric Provider class that computes this Metric declares an additonal metric:

                    .. code-block:: python

                        function_metric_name = "column_values.z_score"

                    The class implements methods decorated with ``@column_function_partial`` to compute the Z score for every row for each backend.

                    Consult the following files for the details of this pattern:

                        * great_expectations/great_expectations/expectations/core/expect_column_value_z_scores_to_be_less_than.py

>                         * great_expectations/great_expectations/expectations/metrics/column_map_metrics/column_values_z_score.py

Reviewer: <3
